src/appengine/handlers/cron/triage.py

```diff
@@ -250,11 +257,105 @@ def _check_and_update_similar_bug(testcase, issue_tracker):
   return False
 
 
-def _file_issue(testcase, issue_tracker):
+def _query_job_bugs_filed_count(job_type, valid_timestamp):
+  """Gets the number of bugs that have been filed for a given job
+  within a given time period."""
+  return data_types.FiledBug.query(
+      data_types.FiledBug.job_type == job_type and
+      data_types.FiledBug.timestamp >= valid_timestamp).count()
+
+
+def _query_project_bugs_filed_count(project_name, valid_timestamp):
+  """Gets the number of bugs that have been filed for a given project
+  within a given time period."""
+  return data_types.FiledBug.query(
+      data_types.FiledBug.project_name == project_name and
+      data_types.FiledBug.timestamp >= valid_timestamp).count()
+
+
+def _get_job_bugs_filing_max(job_type):
+  """Gets the maximum number of bugs that can be filed for a given job."""
+  if job_type in _bug_filing_max_24_hours_per_job:
+    return _bug_filing_max_24_hours_per_job[job_type]
+
+  max_bugs = None
+  job = data_types.Job.query(data_types.Job.name == job_type).get()
+  if job and 'BUG_FILING_MAX_24_HOURS_PER_JOB' in job.get_environment():
+    try:
+      max_bugs = int(job.get_environment()['BUG_FILING_MAX_24_HOURS_PER_JOB'])
+    except Exception:
+      logs.log_error(
+          'Invalid environment value of \'BUG_FILING_MAX_24_HOURS_PER_JOB\' '
+          f'for job type {job_type}.')
+
+  _bug_filing_max_24_hours_per_job[job_type] = max_bugs
+  return max_bugs
+
+
+@lru_cache(1)
+def _get_project_bugs_filing_max():
+  """Gets the maximum number of bugs that can be filed per project."""
+  config = local_config.IssueTrackerConfig()
+  try:
+    return int(
+        config.get('BUG_FILING_MAX_24_HOURS_PER_PROJECT',
+                   BUG_FILING_MAX_24_HOURS_PER_PROJECT_DEFAULT))
+  except:
+    logs.log_error(
+        'Invalid config value of \'BUG_FILING_MAX_24_HOURS_PER_PROJECT\'')
+
+  return BUG_FILING_MAX_24_HOURS_PER_PROJECT_DEFAULT
+
+
+def _throttle_bug(testcase, bug_filed_24_hours_per_job,
+                  bug_filed_24_hours_per_project):
+  """Returns whether the current bug needs to be throttled."""
+  job_bugs_filing_max = _get_job_bugs_filing_max(testcase.job_type)
+  valid_timestamp = datetime.datetime.now() - datetime.timedelta(hours=24)
```

Comment:
Rather than re-computing the valid_timestamp for every individual testcase, we should just use the start time of the triage job. 

That way, we only compute the window at most once per job/project.